gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/SpecStoreChangeMonitor.java

```diff
@@ -0,0 +1,159 @@
+ * a connector between the API and execution layers of GaaS.
+ */
+@Slf4j
+public class SpecStoreChangeMonitor extends HighLevelConsumer {
+  public static final String SPEC_STORE_CHANGE_MONITOR_PREFIX = "specStoreChangeMonitor";
+  static final String SPEC_STORE_CHANGE_MONITOR_TOPIC_KEY = "topic";
+  static final String SPEC_STORE_CHANGE_MONITOR_NUM_THREADS_KEY = "numThreads";
+
+  // Metrics
+  private Meter successfullyAddedSpecs;
+  private Meter failedAddedSpecs;
+  private Meter deletedSpecs;
+  private Meter unexpectedErrors;
+
+  protected CacheLoader<String, String> cacheLoader = new CacheLoader<String, String>() {
+    @Override
+    public String load(String key) throws Exception {
+      return key;
+    }
+  };
+
+  protected LoadingCache<String, String>
+      specChangesSeenCache = CacheBuilder.newBuilder().expireAfterWrite(10, TimeUnit.MINUTES).build(cacheLoader);
+
+  @Inject
+  protected FlowCatalog flowCatalog;
+
+  @Inject
+  protected GobblinServiceJobScheduler scheduler;
+
+  @Inject
+  public SpecStoreChangeMonitor(String topic, Config config, int numThreads) {
+    super(topic, config, numThreads);
+  }
+
+  @Override
+  /*
+  Note that although this class is multi-threaded and will call this message for multiple threads (each having a queue
+  associated with it), a given message itself will be partitioned and assigned to only one queue.
+   */
+  protected void processMessage(DecodeableKafkaRecord message) {
+    String specUri = (String) message.getKey();
+    SpecStoreChangeEvent value = (SpecStoreChangeEvent) message.getValue();
+
+    Long timestamp = value.getTimestamp();
+    String operation = value.getOperationType().name();
+    log.info("Processing message with specUri is {} timestamp is {} operation is {}", specUri, timestamp, operation);
+
+    // If we've already processed a message with this timestamp and spec uri before then skip duplicate message
+    String changeIdentifier = timestamp.toString() + specUri;
+    if (specChangesSeenCache.getIfPresent(changeIdentifier) != null) {
+      return;
+    }
+
+    // If event is a heartbeat type then log it and skip processing
+    if (operation == "HEARTBEAT") {
+      log.info("Received heartbeat message from time {}", timestamp);
+      return;
+    }
+
+    Spec spec;
+    URI specAsUri = null;
+
+    try {
+      specAsUri = new URI(specUri);
+    } catch (URISyntaxException e) {
+      log.warn("Could not create URI object for specUri {} due to error {}", specUri, e.getMessage());
+    }
+
+    spec = (operation != "DELETE") ? this.flowCatalog.getSpecWrapper(specAsUri) : null;
+
+    // The monitor should continue to process messages regardless of failures with individual messages, instead we use
+    // metrics to keep track of failure to process certain SpecStoreChange events
+    try {
+      // Call respective action for the type of change received
+      AddSpecResponse response;
+      if (operation == "CREATE" || operation == "INSERT") {
```

Comment:
What's the difference?

Reply:
ah you're right, keep getting the Brooklin [opcodes](https://jarvis.corp.linkedin.com/codesearch/result/?name=BrooklinEnvelopeMetadataConstants.java&path=datastream%2Fdatastream%2Fdatastream-common%2Fsrc%2Fmain%2Fjava%2Fcom%2Flinkedin%2Fdatastream%2Fcommon&reponame=datastream%2Fdatastream) mixed up with the actions we take. 

INSERT/UPDATE -> create/update spec -> add spec
DELETE -> deletespec